README
NAME
    POE::Component::Win32::ChangeNotify - A POE wrapper around
    Win32::ChangeNotify.

VERSION
    version 1.22

SYNOPSIS
       use strict;
       use POE;
       use POE::Component::Win32::ChangeNotify;

       my $poco = POE::Component::Win32::ChangeNotify->spawn( alias => 'blah' );

       POE::Session->create(
            package_states => [ 
                    'main' => [ qw(_start notification) ],
            ],
       );

       $poe_kernel->run();
       exit 0;

       sub _start {
         my ($kernel,$heap) = @_[KERNEL,HEAP];

         $kernel->post( 'blah' => monitor => 
         {
            'path' => '.',
            'event' => 'notification',
            'filter' => 'ATTRIBUTES DIR_NAME FILE_NAME LAST_WRITE SECURITY SIZE',
            'subtree' => 1,
         } );

         undef;
       }

       sub notification {
         my ($kernel,$hashref) = @_[KERNEL,ARG0];

         if ( $hashref->{error} ) {
            print STDERR $hashref->{error} . "\n";
         } else {
            print STDOUT "Something changed in " . $hashref->{path} . "\n";
         }
         $kernel->post( 'blah' => 'shutdown' );
         undef;
       }

DESCRIPTION
    POE::Component::Win32::ChangeNotify is a POE wrapper around
    Win32::ChangeNotify that provides non-blocking change notify events to
    your POE sessions.

METHODS
    "spawn"
        Takes a number of arguments, all of which are optional.

          'alias', the kernel alias to bless the component with; 
          'options', a hashref of POE::Session options that are passed to the component's 
                     session creator.

    "session_id"
        Takes no arguments, returns the POE::Session ID of the component.
        Useful if you don't want to use aliases.

INPUT
    These are the events that the component will accept.

    "monitor"
        Starts monitoring the specified path for the specified types of
        changes.

        Accepts one argument, a hashref containing the following keys:

          'path', the filesystem path to monitor, mandatory; 
          'event', the event handler to post results back to, mandatory.
          'filter', a string containing whitespace or '|' separated notification flags, 
                    see Win32::ChangeNotify for details, defaults to all notification flags; 
          'subtree', set this to true value to monitor all subdirectories under 'path'; 

             $kernel->post( 'blah' => monitor => 
             {
                'path' => '.',
                'event' => 'notification',
                'filter' => 'ATTRIBUTES DIR_NAME FILE_NAME LAST_WRITE SECURITY SIZE',
                'subtree' => 1,
             } );

    "unmonitor"
        Stops monitoring the specified path.

        Accepts one mandatory argument, 'path', the filesystem path to stop
        monitoring.

           $kernel->post( 'blah' => unmonitor => path => '.' );

    "shutdown"
        Has no arguments. Shuts down the component gracefully. All monitored
        paths will be closed.

OUTPUT
    Each event sent by the component has a hashref as ARG0. This is the
    hashref that was passed to the component with monitor(). The key
    'result' will be true if a change has occurred. The key 'error' will be
    set if an error occurred during the setup of the Win32::ChangeNotify
    object or when trying to call 'wait' on the object.

CAVEATS
    This module will only work on Win32. But you guessed that already :)

SEE ALSO
    Win32::ChangeNotify

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Chris Williams.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2008-11-01 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 1.22 at 2013-10-31 23:27:53 +0000
-----------------------------------------

  Change: 392c7ba9d61de9e5ceaf62064d1998725015d175
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-10-31 23:27:53 +0000

    Update to Dist::Zilla distribution 

-----------------------------------------
version 1.20 at 2009-06-02 13:18:11 +0000
-----------------------------------------

  Change: a6b54d5ba576ccfb6a0b02c59863df0cbfeba0d8
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-06-02 14:18:11 +0000

    Update Changes 

  Change: 4c138db49950b5dbb74e0db76f3d54e8db7e8eeb
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-06-02 14:10:57 +0000

    Update the error testcase slightly 

  Change: 3fb2b44be50d16c0ad93c02b97534502a13422cc
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-06-02 13:11:45 +0000

    Added an error testcase 

  Change: 45edc339e419aa53ca5e0224c073836fcf848174
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-06-02 10:48:26 +0000

    Update Makefile.PL meta and add additional diagnostics when
    Win32::ChangeNotify errors. 

-----------------------------------------
version 1.18 at 2009-04-24 10:47:18 +0000
-----------------------------------------

  Change: 60fe6f86f6cc59fc39c747abf1af097f8fcfa539
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-24 11:47:18 +0000

    Switched to git repository. Tidied up the POD 

  Change: c97a985a0c60dfecde65370bdf24e717cb950434
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-23 13:07:26 +0000

    Tidied up the distribution. Added M::I::AssertOS support. Set
    repository in META.yml 

  Change: b81bb9f19251626ca738226a702b6c62abed7925
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:41:41 +0000

    Removed the lower-case readme file 

-----------------------------------------
version 1.16 at 2009-04-22 22:40:21 +0000
-----------------------------------------

  Change: 4f2e4cf2a3f00762065ed8001c4c1ccf8c58401d
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:40:21 +0000

    Added license information 

-----------------------------------------
version 1.14 at 2009-04-22 22:38:46 +0000
-----------------------------------------

  Change: c57fd4394813ce65d52fd4b99341ce3a79f3ff68
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:38:46 +0000

    Author tests to xt/ 

-----------------------------------------
version 1.12 at 2009-04-22 22:36:07 +0000
-----------------------------------------

  Change: 50b65197db90bc658eabb9226ff88f1654045a0d
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:36:07 +0000

    Oops forgot to no_index 'mylib'. 

-----------------------------------------
version 1.10 at 2009-04-22 22:31:48 +0000
-----------------------------------------

  Change: 04e9a5064bd72ed254aae4450555b4fa776c7cf5
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:31:48 +0000

    Moved Devel::AssertOS to mylib/ to avoid Module::Install trampling on
    it. 

-----------------------------------------
version 1.08 at 2009-04-22 22:29:26 +0000
-----------------------------------------

  Change: 75d9e098261ffe5e3a88536e91e65548641c3b14
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:29:26 +0000

    Added Devel::AssertOS checks to Makefile.PL 

-----------------------------------------
version 1.07 at 2009-04-22 22:27:48 +0000
-----------------------------------------

  Change: 6b683914abc677e03bce83181e8712aea4e8f206
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:27:48 +0000

    Tidied up test names 

-----------------------------------------
version 1.06 at 2009-04-22 22:26:09 +0000
-----------------------------------------

  Change: 41da79178ebce17369df6794d1eb7dba8d6777b5
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:26:09 +0000

    Updated Module::Install to 0.68 

-----------------------------------------
version 1.05 at 2009-04-22 22:24:52 +0000
-----------------------------------------

  Change: 37c1eee6ea627a9e3539ff19e06454d387d6a560
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:24:52 +0000

    Fixes for build_requires and abstract_from in Makefile.PL 

-----------------------------------------
version 1.04 at 2009-04-22 22:23:13 +0000
-----------------------------------------

  Change: 27fca11aa06da27367e6e9b97ab5bc06cc25509b
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:23:13 +0000

    Makefile.PL is now all M::I. Changed to sig_child() semantics. 

-----------------------------------------
version 1.03 at 2009-04-22 22:20:59 +0000
-----------------------------------------

  Change: 53f6269410b2a85e248f648f3cf93544e57a3fb6
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:20:59 +0000

    Switched to Module::Install 

-----------------------------------------
version 1.02 at 2009-04-22 22:19:23 +0000
-----------------------------------------

  Change: e575853e78d5140c4ef8b26f303a90bdf9a71d33
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:19:23 +0000

    - Added Changes file. - Added pod and pod coverage tests. - Added
    examples/ folder. - Fixed use strict. 

-----------------------------------------
version 1.01 at 2009-04-22 22:18:07 +0000
-----------------------------------------

  Change: e3339bc8200cb4b4530402ceeb3ae580a57e33d7
  Author: Chris Williams <chris@bingosnet.co.uk>
  Date : 2009-04-22 23:18:07 +0000

    Import 1.01 

-----------------------------------------
version 1.00 at 2009-04-22 22:14:48 +0000
-----------------------------------------

  Change: 080890229b8a7cf36b15fd4feab0482bdaf91bb7

Makefile.PL
use lib 'inc';
use Devel::AssertOS qw[MSWin32 Cygwin];

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A POE wrapper around Win32::ChangeNotify.\r",
  "AUTHOR" => "Chris Williams <chris\@bingosnet.co.uk>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "POE-Component-Win32-ChangeNotify",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "POE::Component::Win32::ChangeNotify",
  "PREREQ_PM" => {
    "POE" => "0.38",
    "POE::Filter::Line" => 0,
    "POE::Filter::Reference" => 0,
    "POE::Wheel::Run" => 0,
    "Win32::ChangeNotify" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.22",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





inc/Devel/CheckOS.pm
# $Id: CheckOS.pm,v 1.32 2008/11/11 23:49:49 drhyde Exp $

package #
Devel::CheckOS;

use strict;
use Exporter;

use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);

$VERSION = '1.6';

# localising prevents the warningness leaking out of this module
local $^W = 1;    # use warnings is a 5.6-ism

@ISA = qw(Exporter);
@EXPORT_OK = qw(os_is os_isnt die_if_os_is die_if_os_isnt die_unsupported list_platforms list_family_members);
%EXPORT_TAGS = (
    all      => \@EXPORT_OK,
    booleans => [qw(os_is os_isnt die_unsupported)],
    fatal    => [qw(die_if_os_is die_if_os_isnt)]
);

=head1 NAME

Devel::CheckOS - check what OS we're running on

=head1 DESCRIPTION

Devel::CheckOS provides a more friendly interface to $^O, and also lets
you check for various OS "families" such as "Unix", which includes things
like Linux, Solaris, AIX etc.

=head1 SYNOPSIS

    use Devel::CheckOS qw(os_is);
    print "Hey, I know this, it's a Unix system\n" if(os_is('Unix'));

    print "You've got Linux 2.6\n" if(os_is('Linux::v2_6'));

=head1 USING IT IN Makefile.PL or Build.PL

If you want to use this from Makefile.PL or Build.PL, do
not simply copy the module into your distribution as this may cause
problems when PAUSE and search.cpan.org index the distro.  Instead, use
the use-devel-assertos script.

=head1 FUNCTIONS

Devel::CheckOS implements the following functions, which load subsidiary
OS-specific modules on demand to do the real work.  They can be exported
by listing their names after C<use Devel::CheckOS>.  You can also export
groups of functions thus:

    use Devel::CheckOS qw(:booleans); # export the boolean functions
                                      # and 'die_unsupported'
    
    use Devel::CheckOS qw(:fatal);    # export those that die on no match

    use Devel::CheckOS qw(:all);      # export everything

=head2 Boolean functions

=head3 os_is

Takes a list of OS names.  If the current platform matches any of them,
it returns true, otherwise it returns false.  The names can be a mixture
of OSes and OS families, eg ...

    os_is(qw(Unix VMS)); # Unix is a family, VMS is an OS

=cut

sub os_is {
    my @targets = @_;
    foreach my $target (@targets) {
        die("Devel::CheckOS: $target isn't a legal OS name\n")
            unless($target =~ /^\w+(::\w+)*$/);
        eval "use Devel::AssertOS::$target";
        if(!$@) {
            no strict 'refs';
            return 1 if(&{"Devel::AssertOS::${target}::os_is"}());
        }
    }
    return 0;
}

=head3 os_isnt

If the current platform matches any of the parameters it returns false,
otherwise it returns true.

=cut

sub os_isnt {
    my @targets = @_;
    foreach my $target (@targets) {
        return 0 if(os_is($target));
    }
    return 1;
}

=head2 Fatal functions

=head3 die_if_os_isnt

As C<os_is()>, except that it dies instead of returning false.  The die()
message matches what the CPAN-testers look for to determine if a module
doesn't support a particular platform.

=cut

sub die_if_os_isnt {
    os_is(@_) ? 1 : die_unsupported();
}

=head3 die_if_os_is

As C<os_isnt()>, except that it dies instead of returning false.

=cut

sub die_if_os_is {
    os_isnt(@_) ? 1 : die_unsupported();
}

=head2 And some utility functions ...

=head3 die_unsupported

This function simply dies with the message "OS unsupported", which is what
the CPAN testers look for to figure out whether a platform is supported or
not.

=cut

sub die_unsupported { die("OS unsupported\n"); }

=head3 list_platforms

When called in list context,
return a list of all the platforms for which the corresponding
Devel::AssertOS::* module is available.  This includes both OSes and OS
families, and both those bundled with this module and any third-party
add-ons you have installed.

In scalar context, returns a hashref keyed by platform with the filename
of the most recent version of the supporting module that is available to you.
This is to make sure that the use-devel-assertos script Does The Right Thing
in the case where you have installed the module in one version of perl, then
upgraded perl, and installed it again in the new version.  Sometimes the old
version of perl and all its modules will still be hanging around and perl
"helpfully" includes the old perl's search path in its own.

Unfortunately, on some platforms this list may have file case
broken.  eg, some platforms might return 'freebsd' instead of 'FreeBSD'.
This is because they have case-insensitive filesystems so things
should Just Work anyway.

=cut

my ($re_Devel, $re_AssertOS);

sub list_platforms {
    eval " # only load these if needed
        use File::Find::Rule;
        use File::Spec;
    ";
    
    die($@) if($@);
    if (!$re_Devel) {
        my $case_flag = File::Spec->case_tolerant ? '(?i)' : '';
        $re_Devel    = qr/$case_flag ^Devel$/x;
        $re_AssertOS = qr/$case_flag ^AssertOS$/x;
    }

    # sort by mtime, so oldest last
    my @modules = sort {
        (stat($a->{file}))[9] <=> (stat($b->{file}))[9]
    } map {
        my (undef, $dir_part, $file_part) = File::Spec->splitpath($_);
        $file_part =~ s/\.pm$//;
        my (@dirs) = grep {+length} File::Spec->splitdir($dir_part);
        foreach my $i (reverse 1..$#dirs) {
            next unless $dirs[$i] =~ $re_AssertOS
                && $dirs[$i - 1] =~ $re_Devel;
            splice @dirs, 0, $i + 1;
            last;
        }
        {
            module => join('::', @dirs, $file_part),
            file   => File::Spec->canonpath($_)
        }
    } File::Find::Rule->file()->name('*.pm')->in(
        grep { -d }
        map { File::Spec->catdir($_, qw(Devel AssertOS)) }
        @INC
    );

    my %modules = map {
        $_->{module} => $_->{file}
    } @modules;

    if(wantarray()) {
        return sort keys %modules;
    } else {
        return \%modules;
    }
}

=head3 list_family_members

Takes the name of an OS 'family' and returns a list of all its members.
In list context, you get a list, in scalar context you get an arrayref.

If called on something that isn't a family, you get an empty list (or
a ref to an empty array).

=cut

sub list_family_members {
    my $family = shift() ||
        die(__PACKAGE__."::list_family_members needs a parameter\n");

    # this will die if it's the wrong OS, but the module is loaded ...
    eval qq{use Devel::AssertOS::$family};
    # ... so we can now query it
    my @members = eval qq{
        no strict 'refs';
	&{"Devel::AssertOS::${family}::matches"}()
    };
    return wantarray() ? @members : \@members;
}

=head1 PLATFORMS SUPPORTED

To see the list of platforms for which information is available, run this:

    perl -MDevel::CheckOS -e 'print join(", ", Devel::CheckOS::list_platforms())'

Note that capitalisation is important.  These are the names of the
underlying Devel::AssertOS::* modules
which do the actual platform detection, so they have to
be 'legal' filenames and module names, which unfortunately precludes
funny characters, so platforms like OS/2 are mis-spelt deliberately.
Sorry.

Also be aware that not all of them have been properly tested.  I don't
have access to most of them and have had to work from information
gleaned from L<perlport> and a few other places.  For a complete list of
OS families, see L<Devel::CheckOS::Families>.

If you want to add your own OSes or families, see L<Devel::AssertOS::Extending>
and please feel free to upload the results to the CPAN.

=head1 BUGS and FEEDBACK

I welcome feedback about my code, including constructive criticism.
Bug reports should be made using L<http://rt.cpan.org/> or by email.

You will need to include in your bug report the exact value of $^O, what
the OS is called (eg Windows Vista 64 bit Ultimate Home Edition), and,
if relevant, what "OS family" it should be in and who wrote it.

If you are feeling particularly generous you can encourage me in my
open source endeavours by buying me something from my wishlist:
  L<http://www.cantrell.org.uk/david/wishlist/>

=head1 SEE ALSO

$^O in L<perlvar>

L<perlport>

L<Devel::AssertOS>

L<Devel::AssertOS::Extending>

L<Probe::Perl>

The use-devel-assertos script

=head1 AUTHOR

David Cantrell E<lt>F<david@cantrell.org.uk>E<gt>

Thanks to David Golden for the name and ideas about the interface, and
to the cpan-testers-discuss mailing list for prompting me to write it
in the first place.

Thanks to Ken Williams, from whose L<Module::Build> I lifted some of the
information about what should be in the Unix family.

Thanks to Billy Abbott for finding some bugs for me on VMS.

Thanks to Matt Kraai for information about QNX.

Thanks to Kenichi Ishigaki and Gabor Szabo for reporting a bug on Windows,
and to the former for providing a patch.

=head1 CVS

L<http://drhyde.cvs.sourceforge.net/drhyde/perlmodules/Devel-CheckOS/>

=head1 COPYRIGHT and LICENCE

Copyright 2007 David Cantrell

This software is free-as-in-speech software, and may be used, distributed, and modified under the terms of either the GNU General Public Licence version 2 or the Artistic Licence. It's up to you which one you use. The full text of the licences can be found in the files GPL2.txt and ARTISTIC.txt, respectively.

=head1 CONSPIRACY

This module is also free-as-in-mason software.

=cut

1;


inc/Module/Install/AssertOS.pm
#line 1
package Module::Install::AssertOS;

use strict;
use warnings;
use base qw(Module::Install::Base);
use File::Spec;
use vars qw($VERSION);

$VERSION = '0.04';

sub assertos {
  my $self = shift;
  my @oses = @_;
  return unless scalar @oses;

  unless ( $Module::Install::AUTHOR ) {
     require Devel::AssertOS;
     Devel::AssertOS->import( @oses );
     return;
  }

  require Data::Compare;

  foreach my $os (@oses) {
    my $oldinc = { map { $_ => $INC{$_} } keys %INC }; # clone
    eval "use Devel::AssertOS qw($os)";
    if(Data::Compare::Compare(\%INC, $oldinc)) {
        print STDERR "Couldn't find a module for $os\n";
        exit(1);
    }
  }
  my @modulefiles = keys %{{map { $_ => $INC{$_} } grep { /Devel/i && /(Check|Assert)OS/i } keys %INC}};

  mkdir 'inc';
  mkdir 'inc/Devel';
  mkdir 'inc/Devel/AssertOS';
  print "Extra directories created under inc/\n";

  foreach my $modulefile (@modulefiles) {
    my $fullfilename = '';
    SEARCHINC: foreach (@INC) {
        if(-e File::Spec->catfile($_, $modulefile)) {
            $fullfilename = File::Spec->catfile($_, $modulefile);
            last SEARCHINC;
        }
    }
    die("Can't find a file for $modulefile\n") unless(-e $fullfilename);

    (my $module = join('::', split(/\W+/, $modulefile))) =~ s/::pm/.pm/;
    my @dircomponents = ('inc', (split(/::/, $module)));
    my $file = pop @dircomponents;

    mkdir File::Spec->catdir(@dircomponents);
    
    open(PM, $fullfilename) ||
        die("Can't read $fullfilename: $!");
    my $lsep = $/;
    $/ = undef;
    (my $pm = <PM>) =~ s/package Devel::/package #\nDevel::/;
    close(PM);
    $/ = $lsep;
    open(PM, '>'.File::Spec->catfile(@dircomponents, $file)) ||
        die("Can't write ".File::Spec->catfile(@dircomponents, $file).": $!");
    print PM $pm;
    print "Copied $fullfilename to\n       ".File::Spec->catfile(@dircomponents, $file)."\n";
    close(PM);

  }
  return 1;
}

'Assert this';

__END__

#line 133


lib/POE/Component/Win32/ChangeNotify.pm
# Author: Chris "BinGOs" Williams
#
# This module may be used, modified, and distributed under the same
# terms as Perl itself. Please see the license that came with your Perl
# distribution for details.
#

package POE::Component::Win32::ChangeNotify;
{
  $POE::Component::Win32::ChangeNotify::VERSION = '1.22';
}

#ABSTRACT: A POE wrapper around Win32::ChangeNotify.

use strict;
use warnings;
use POE 0.38 qw(Wheel::Run Filter::Reference Filter::Line);
use Win32;
use Win32::ChangeNotify;
use Carp qw(carp croak);

sub spawn {
  my $package = shift;
  croak "$package needs an even number of parameters" if @_ & 1;
  my %params = @_;

  $params{ lc $_ } = delete $params{ $_ } for keys %params;
  my $options = delete $params{'options'};

  my $self = bless \%params, $package;

  $self->{session_id} = POE::Session->create(
	  object_states => [
		  $self => [ qw(_start _sig_chld child_closed child_error child_stderr child_stdout shutdown monitor unmonitor) ],
	  ],
	  ( ( defined ( $options ) and ref ( $options ) eq 'HASH' ) ? ( options => $options ) : () ),
  )->ID();

  $self->{FILTERS} = [ qw(ATTRIBUTES DIR_NAME FILE_NAME LAST_WRITE SECURITY SIZE) ];

  return $self;
}

sub _start {
  my ($kernel,$self) = @_[KERNEL,OBJECT];

  $self->{session_id} = $_[SESSION]->ID();

  if ( $self->{alias} ) {
	$kernel->alias_set( $self->{alias} );
  } else {
	$kernel->refcount_increment( $self->{session_id} => __PACKAGE__ );
  }

  #$kernel->sig( 'CHLD' => '_sig_chld' );
  $self->{wheels} = { };
  undef;
}

sub _sig_chld {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  $kernel->sig_handled();
}

sub shutdown {
  my ($kernel,$self) = @_[KERNEL,OBJECT];

  if ( $self->{alias} ) {
	$kernel->alias_remove( $_ ) for $kernel->alias_list();
  } else {
	$kernel->refcount_decrement( $self->{session_id} => __PACKAGE__ );
  }

  # Clean up wheels before we go.
  $self->{_wheel_count} = scalar keys %{ $self->{wheels} };
  foreach my $wheel_id ( keys %{ $self->{wheels} } ) {
	  $self->{wheels}->{ $wheel_id }->{wheel}->kill(9);
	  my $wheel_data = delete $self->{wheels}->{ $wheel_id };
	  $kernel->refcount_decrement( $wheel_data->{sender} => __PACKAGE__ );
  }
  delete $self->{monitored};
  #$kernel->sig( 'CHLD' ) unless $self->{_wheel_count};
  $self->{_shutdown} = 1;
  undef;
}

sub monitor {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  my $sender = $_[SENDER]->ID();

  # Get the arguments
  my $args;
  if (ref($_[ARG0]) eq 'HASH') {
	$args = { %{ $_[ARG0] } };
  } else {
	warn "first parameter must be a ref hash, trying to adjust. "
		."(fix this to get rid of this message)";
	$args = { @_[ARG0 .. $#_ ] };
  }

  $args->{ lc $_ } = delete $args->{$_} for keys %{ $args };
  
  unless ( $args->{path} ) {
	warn "you must supply a path argument, otherwise what's the point";
	return;
  }

  unless ( $args->{event} ) {
	warn "you must supply an event argument, otherwise where do I send the replies to";
	return;
  }

  unless ( $args->{filter} ) {
	  $args->{filter} = join(' ', @{ $self->{FILTERS} } );
  }

  $args->{sender} = $sender;
  
  unless ( defined ( $self->{monitored}->{ $args->{path} } ) ) {
    $args->{always} = 1;
    my $wheel = POE::Wheel::Run->new(
    	Program     => \&_launch_change_notify,
    	ErrorEvent  => 'child_error',             # Event to emit on errors.
    	CloseEvent  => 'child_closed',     # Child closed all output.

    	StdoutEvent => 'child_stdout', # Event to emit with child stdout information.
    	StderrEvent => 'child_stderr', # Event to emit with child stderr information.
    	StdioFilter  => POE::Filter::Reference->new(),
	StderrFilter => POE::Filter::Line->new(),   
    	CloseOnCall => 0,
    );

    if ( $wheel ) {
    	$kernel->refcount_increment( $sender => __PACKAGE__ );
    	$wheel->put( $args );
    	$self->{wheels}->{ $wheel->ID() }->{wheel} = $wheel;
    	$self->{wheels}->{ $wheel->ID() }->{args} = $args;
    	$self->{wheels}->{ $wheel->ID() }->{sender} = $sender;
	$self->{monitored}->{ $args->{path} } = $wheel->ID();
        $kernel->sig_child( $wheel->PID, '_sig_chld' );
    }
  }

  undef;
}

sub unmonitor {
  my ($kernel,$self,$state) = @_[KERNEL,OBJECT,STATE];
  croak "$state event needs an even number of parameters" if @_ & 1;
  my %params = @_;

  $params{ lc $_ } = delete ( $params{ $_ } ) for keys %params;

  my $path = delete $params{path};

  croak "$state requires a path parameter" unless $path;

  if ( defined ( $self->{monitored}->{ $path } ) ) {
	my $wheel_id = delete $self->{monitored}->{ $path };
	$self->{wheels}->{ $wheel_id }->{wheel}->kill();
	my $wheel_data = delete $self->{wheels}->{ $wheel_id };
	$kernel->refcount_decrement( $wheel_data->{sender} => __PACKAGE__ );
  }

  undef;
}

sub child_closed {
  my ($self,$wheel_id) = @_[OBJECT,ARG0];

  my $wheel_data = delete $self->{wheels}->{ $wheel_id };
  delete $self->{monitored}->{ $wheel_data->{args}->{path} };
  undef;
}

sub child_error {
  my ($self,$wheel_id) = @_[OBJECT,ARG3];

  my $wheel_data = delete $self->{wheels}->{ $wheel_id };
  delete $self->{monitored}->{ $wheel_data->{args}->{path} };
  undef;
}

sub child_stdout {
  my ($kernel,$self,$input,$wheel_id) = @_[KERNEL,OBJECT,ARG0,ARG1];
  delete $input->{always};
  my $event = delete $input->{event};
  my $sender = delete $input->{sender};

  $kernel->post( $sender => $event => $input );
  undef;
}

sub child_stderr {
  my ($kernel,$self,$input,$wheel_id) = @_[KERNEL,OBJECT,ARG0,ARG1];

  undef;
}

sub session_id {
  return $_[0]->{session_id};
}

sub _launch_change_notify {
  binmode(STDIN); binmode(STDOUT); 
  my $raw;
  my $size = 4096;
  my $filter = POE::Filter::Reference->new();

  READ:
  while ( sysread ( STDIN, $raw, $size ) ) {
    my $requests = $filter->get( [ $raw ] );
    _watch_path($filter,$_) for @{ $requests };
  }
}

sub _watch_path {
  my ($filter,$req) = splice @_, 0, 2;

  my $notify = Win32::ChangeNotify->new($req->{path}, $req->{subtree}, $req->{filter});

  unless ( $notify ) {
	delete $req->{result};
	$req->{error} = "Couldn't create Win32::ChangeNotify object: " .
				Win32::FormatMessage( Win32::GetLastError() );
	my $replies = $filter->put( [ $req ] );
	print STDOUT @$replies;
	return 0;
  }

  while ( $req->{always} ) {
	$notify->reset;
	my $result = $notify->wait;

	if ( $result ) {
	  $req->{result} = $result;
	  my $replies = $filter->put( [ $req ] );
	  print STDOUT @$replies;
	} else {
	  delete $req->{result};
	  $req->{error} = "Something bad happened with the Win32::ChangeNotify object";
	  my $replies = $filter->put( [ $req ] );
	  print STDOUT @$replies;
	}
  }
  $notify->close;
  return 1;
}

1;

__END__

=pod

=head1 NAME

POE::Component::Win32::ChangeNotify - A POE wrapper around Win32::ChangeNotify.

=head1 VERSION

version 1.22

=head1 SYNOPSIS

   use strict;
   use POE;
   use POE::Component::Win32::ChangeNotify;

   my $poco = POE::Component::Win32::ChangeNotify->spawn( alias => 'blah' );

   POE::Session->create(
   	package_states => [ 
		'main' => [ qw(_start notification) ],
	],
   );

   $poe_kernel->run();
   exit 0;

   sub _start {
     my ($kernel,$heap) = @_[KERNEL,HEAP];

     $kernel->post( 'blah' => monitor => 
     {
     	'path' => '.',
	'event' => 'notification',
	'filter' => 'ATTRIBUTES DIR_NAME FILE_NAME LAST_WRITE SECURITY SIZE',
	'subtree' => 1,
     } );

     undef;
   }

   sub notification {
     my ($kernel,$hashref) = @_[KERNEL,ARG0];

     if ( $hashref->{error} ) {
     	print STDERR $hashref->{error} . "\n";
     } else {
     	print STDOUT "Something changed in " . $hashref->{path} . "\n";
     }
     $kernel->post( 'blah' => 'shutdown' );
     undef;
   }

=head1 DESCRIPTION

POE::Component::Win32::ChangeNotify is a POE wrapper around L<Win32::ChangeNotify> that provides non-blocking change notify events to your POE sessions.

=head1 METHODS

=over

=item C<spawn>

Takes a number of arguments, all of which are optional.

  'alias', the kernel alias to bless the component with; 
  'options', a hashref of POE::Session options that are passed to the component's 
             session creator.

=item C<session_id>

Takes no arguments, returns the L<POE::Session> ID of the component. Useful if you don't want to use
aliases.

=back

=head1 INPUT

These are the events that the component will accept.

=over

=item C<monitor>

Starts monitoring the specified path for the specified types of changes.

Accepts one argument, a hashref containing the following keys: 

  'path', the filesystem path to monitor, mandatory; 
  'event', the event handler to post results back to, mandatory.
  'filter', a string containing whitespace or '|' separated notification flags, 
            see Win32::ChangeNotify for details, defaults to all notification flags; 
  'subtree', set this to true value to monitor all subdirectories under 'path'; 

     $kernel->post( 'blah' => monitor => 
     {
     	'path' => '.',
	'event' => 'notification',
	'filter' => 'ATTRIBUTES DIR_NAME FILE_NAME LAST_WRITE SECURITY SIZE',
	'subtree' => 1,
     } );

=item C<unmonitor>

Stops monitoring the specified path.

Accepts one mandatory argument, 'path', the filesystem path to stop monitoring.

   $kernel->post( 'blah' => unmonitor => path => '.' );

=item C<shutdown>

Has no arguments. Shuts down the component gracefully. All monitored paths will be closed.

=back

=head1 OUTPUT

Each event sent by the component has a hashref as ARG0. This is the hashref that was passed to the component with monitor(). The key 'result' will be true if a change has occurred. The key 'error' will be set if an error occurred during the setup of the Win32::ChangeNotify object or when trying to call 'wait' on the object. 

=head1 CAVEATS

This module will only work on Win32. But you guessed that already :)

=head1 SEE ALSO

L<Win32::ChangeNotify>

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
